File: javascript/programmers/javascript/level3/자물쇠와열쇠.js
const key = [
  [0, 0, 0],
  [1, 0, 0],
  [0, 1, 1],
];
const lock = [
  [1, 1, 1],
  [1, 1, 0],
  [1, 0, 1],
];

const rotationKey = (key) => {
  const len = key.length;
  const ret = Array.from(Array(len), () => Array(len).fill(null));
  for (let i = 0; i < len; ++i) {
    for (let j = 0; j < len; ++j) {
      ret[i][j] = key[len - j - 1][i];
    }
  }
  return ret;
};

const isAnswer = (newLock, len) => {
  for (let i = len; i < len * 2; i++) {
    for (let j = len; j < len * 2; j++) {
      if (newLock[i][j] !== 1) {
        return false;
      }
    }
  }
  return true;
};
const solution = (key, lock) => {
  let answer = true;
  const len = lock.length;
  const arr = Array.from(Array(len * 3), () => Array(len * 3).fill(null));

  for (let i = len; i < len * 2; i++) {
    for (let j = len; j < len * 2; j++) {
      arr[i][j] = lock[i - len][j - len];
    }
  }
  //console.log(arr);

  //키를 회전 시키면서 탐색
  for (let i = 0; i < 4; i++) {
    key = rotationKey(key, i);
    //키를 이동시키면서 탐색
    for (let j = 0; j <= arr.length - key.length; j++) {
      for (let k = 0; k <= arr[0].length - key[0].length; k++) {
        const newLock = arr.map(function (arr) {
          return arr.slice();
        });
        for (let m = 0; m < key.length; m++) {
          for (let n = 0; n < key.length; n++) {
            if (newLock[j + m][k + n] === 1 && key[m][n] === 1) {
              newLock[j + m][k + n] = 2;
            } else if (newLock[j + m][k + n] === 1 && key[m][n] === 0) {
              newLock[j + m][k + n] = 1;
            } else {
              newLock[j + m][k + n] = key[m][n];
            }
          }
        }
        //console.log(`새로운 Arr`);
        // console.log(newLock);
        if (isAnswer(newLock, len)) {
          //console.log(newLock);
          return true;
        }
      }
    }
  }
  return false;
};

console.log(solution(key, lock));

File: python/programmers/level3/자물쇠와 열쇠.py
def attach(x, y, M, key, board):
    for i in range(M):
        for j in range(M):
            board[x + i][y + j] += key[i][j]


def detach(x, y, M, key, board):
    for i in range(M):
        for j in range(M):
            board[x + i][y + j] -= key[i][j]


def rotate(arr, d):
    l = len(arr)
    ret = [[0] * l for _ in range(l)]

    if d % 4 == 1:
        for i in range(l):
            for j in range(l):
                ret[j][l - 1 - i] = arr[i][j]
    elif d % 4 == 2:
        for i in range(l):
            for j in range(l):
                ret[l - 1 - i][l - 1 - j] = arr[i][j]
    elif d % 4 == 3:
        for i in range(l):
            for j in range(l):
                ret[l - 1 - j][i] = arr[i][j]
    else:
        for i in range(l):
            for j in range(l):
                ret[i][j] = arr[i][j]

    return ret


def isAnswer(newLock, length):
    for i in range(length, length * 2):
        for j in range(length, length * 2):
            if newLock[i][j] != 1:
                return False
    return True


def solution(key, lock):
    lockLen = len(lock)
    keyLen = len(key)
    newLocklength = lockLen * 3
    arr = [[0] * newLocklength for _ in range(newLocklength)]
    for i in range(lockLen, lockLen * 2):
        for j in range(lockLen, lockLen * 2):
            arr[i][j] = lock[i - lockLen][j - lockLen]

    for i in range(4):
        rotateKey = rotate(key, i)
        for j in range(newLocklength - keyLen):
            for k in range(newLocklength - keyLen):
                attach(j, k, keyLen, rotateKey, arr)
                if isAnswer(arr, lockLen):
                    return True
                detach(j, k, keyLen, rotateKey, arr)
    return False


print(solution([[0, 0, 0], [1, 0, 0], [0, 1, 1]], [[1, 1, 1], [1, 1, 0], [1, 0, 1]]))
